engines/sci/sound/audio32.cpp

```diff
@@ -0,0 +1,920 @@
+
+	if (channelIndex == kNoExistingChannel || oldNumChannels == 0) {
+		return 0;
+	}
+
+	if (channelIndex == kAllChannels) {
+		for (int i = 0; i < oldNumChannels; ++i) {
+			freeChannel(i);
+		}
+		_numActiveChannels = 0;
+	} else {
+		freeChannel(channelIndex);
+		--_numActiveChannels;
+		for (int i = channelIndex; i < oldNumChannels - 1; ++i) {
+			_channels[i] = _channels[i + 1];
+			if (i + 1 == _monitoredChannelIndex) {
+				_monitoredChannelIndex = i;
+			}
+		}
+	}
+
+	// NOTE: SSCI stops the DSP interrupt and frees the
+	// global decompression buffer here if there are no
+	// more active channels
+	if (_numActiveChannels == 0) {
+		_mixer->pauseHandle(_handle, true);
+	}
+
+	return oldNumChannels;
+}
+
+int16 Audio32::getPosition(const int16 channelIndex) const {
+	Common::StackLock lock(_mutex);
+	if (channelIndex == kNoExistingChannel || _numActiveChannels == 0) {
+		return -1;
+	}
+
+	// NOTE: SSCI treats this as an unsigned short except for
+	// when the value is 65535, then it treats it as signed
+	int position = -1;
+	const uint32 now = g_sci->getTickCount();
+
+	// NOTE: The original engine also queried the audio driver to see whether
+	// it thought that there was audio playback occurring via driver opcode 9
+	if (channelIndex == kAllChannels) {
+		if (_pausedAtTick) {
+			position = _pausedAtTick - _startedAtTick;
+		} else {
+			position = now - _startedAtTick;
+		}
+	} else {
+		const AudioChannel &channel = getChannel(channelIndex);
+
+		if (channel.pausedAtTick) {
+			position = channel.pausedAtTick - channel.startedAtTick;
+		} else if (_pausedAtTick) {
+			position = _pausedAtTick - channel.startedAtTick;
+		} else {
+			position = now - channel.startedAtTick;
+		}
+	}
+
+	return MIN(position, 65534);
+}
+
+void Audio32::setLoop(const int16 channelIndex, const bool loop) {
+	Common::StackLock lock(_mutex);
+
+	if (channelIndex < 0 || channelIndex >= _numActiveChannels) {
+		return;
+	}
+
+	AudioChannel &channel = getChannel(channelIndex);
+	channel.loop = loop;
+}
+
+reg_t Audio32::kernelPlay(const bool autoPlay, const int argc, const reg_t *const argv) {
+	if (argc == 0) {
+		return make_reg(0, getNumActiveChannels());
+	}
+
+	const int16 channelIndex = findChannelByArgs(argc, argv, 0, NULL_REG);
+	ResourceId resourceId;
+	bool loop;
+	int16 volume;
+	bool monitor = false;
+	reg_t soundNode;
+
+	if (argc >= 5) {
+		resourceId = ResourceId(kResourceTypeAudio36, argv[0].toUint16(), argv[1].toUint16(), argv[2].toUint16(), argv[3].toUint16(), argv[4].toUint16());
+
+		if (argc < 6 || argv[5].toSint16() == 1) {
+			loop = false;
+		} else {
+			// NOTE: Uses -1 for infinite loop. Presumably the
+			// engine was supposed to allow counter loops at one
+			// point, but ended up only using loop as a boolean.
+			loop = (bool)argv[5].toSint16();
+		}
+
+		if (argc < 7 || argv[6].toSint16() < 0 || argv[6].toSint16() > Audio32::kMaxVolume) {
+			volume = Audio32::kMaxVolume;
+
+			if (argc >= 7) {
+				monitor = true;
```

Comment:
I'm confused here. Is there a specific call that marks a channel to be monitored for signals? It seems like the 7th argument is the volume, and you're checking for all values that are out of the ordinary? Isn't there a specific mark, like -1?

Also, you are checking for all calls with 7 or more arguments... Shouldn't this be exactly 7 arguments?